### DISCUSSION TITLE
USB to iic

### DISCUSSION THREAD

**Discussion by fakircurry** [robhagemans/pcbasic#175] [General]
Hello,
My name is Tomas.
I am trying to use a chip (FTDI FT232R) that converts usb to iic. The intention is to use iic bus to control several pcf8574 (iic to parallell in/out). I have now tried  a lot of combinations of comands in PC-Basic but it will not work. The chip is available on a small board from Robot Electronics in UK, there is also a link to FTDI and all documentation. Can somebody help me to figure out if it is possible to do it, and in that case what to do.

**Comment by Marrin** [comment]
As far as I can see that FTDI chip doesn't ”speak” i²c but RS232 & Co.  PC-BASIC can access [serial ports](https://en.wikipedia.org/wiki/Serial_port) and that chip is accessible as a serial port, so if that is the interface you need it may be possible to use PC-BASIC.

**Reply by fakircurry** [reply]
Thank you for the response. You are right about the chip, it has rs232, but on the module it is another chip that convert the communication to iic. I found one thing that could be a problem. The specification says that the baudrate should be set to 19200 (I suppose it is between the chips) and PC-Basic maximum setting is 9600 baud.

**Comment by fakircurry** [comment]
I am a little confused with "open" command. Is it setting the baudrate on the simulated com port which goes to the USB port or is it setting the baudrate on the external device?

here is my simple test software.
10 OPEN "com1:9600,N,8,1,RS,CS0,DS0,CD0" AS #1
20 INPUT A
30 IF A=256 THEN 140: REM read
40 REM * write to pcf 8574 *
50 A$=CHR$(&H53)+CHR$(&H40)+CHR$(A)
60 PRINT #1, A$;
70 FOR XX=1 TO 1000:NEXT XX
80 IF EOF(1) THEN 120
90 C$=INPUT$(1,#1)
100 PRINT"ack"
110 PRINT VAL(C$)
120 GOTO 20
130 REM * read *
140 A$=CHR$(&H53)+CHR$(&H41)
150 PRINT #1, A$;
160 FOR XX=1 TO 1000:NEXT XX
170 IF EOF(1) THEN 210
180 B$=INPUT$(1,#1)
190 PRINT "data in"
200 PRINT VAL(B$)
210 GOTO 20

**Comment by fakircurry** [comment]
I found the problem!
There are two chips on the module and communication must be fixed to 19200 baud between them. I measured with oscilloscope and saw the difference when using PC-Basic control and the test-tool . Is it possible to do something from PC-Basic to open the com1 port with 19200 baud and not the maximum 9600?

**Reply by udhos** [reply]
Maybe you could manually tweak ports.py:

https://github.com/robhagemans/pcbasic/blob/master/pcbasic/basic/devices/ports.py#L96

**Comment by fakircurry** [comment]
Hi,
Thank you for the respons.
I have looked in to the code and understand what you mean, but since it was more then 30 years ago I wrote sw code I am not really sure what to do. Should I compile an own version of PC-Basic? I think I need more help.


________________________________
Från: udhos ***@***.***>
Skickat: den 15 januari 2022 02:58
Till: robhagemans/pcbasic ***@***.***>
Kopia: fakircurry ***@***.***>; Author ***@***.***>
Ämne: Re: [robhagemans/pcbasic] USB to iic (Discussion #175)


Maybe you could manually tweak ports.py:

https://github.com/robhagemans/pcbasic/blob/master/pcbasic/basic/devices/ports.py#L96

—
Reply to this email directly, view it on GitHub<https://github.com/robhagemans/pcbasic/discussions/175#discussioncomment-1973202>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AW5X3URZUMCGUGCIASCCU43UWDPF3ANCNFSM5LZLXYNA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

**Comment by robhagemans** [comment]
What's the reason you're trying to use PC-BASIC to address this appliance? Do you have some complex legacy BASIC code that you are trying to reuse? 

If not, it might be much easier to use Python and the PySerial module directly - see https://pyserial.readthedocs.io/en/latest/pyserial.html. These are what PC-BASIC is built on, but PC-BASIC is primarily intended to closely emulate GW-BASIC along with that language's limitations and quirks, such as the limitation of the baudrate to a number of predefined settings with a maximum of 9600.

I think for your purposes using Python and PySerial would lead to much simpler and easier to debug and maintain code - it's certainly cleaner than creating a modified version of PC-BASIC and using that.

**Comment by fakircurry** [comment]
Very good question. The only reason is that I have programmed in Basic earlier and know how it works, it is very easy to change in the sw and run it without compiler. To make a store short. I have built a large model train installation (analouge märklin). For two months ago I bought an Arduino to try what is possible to do with it, now the Arduino is controlling some gadgets. My new idea was to control the whole train installation, more then 100 relays. I found PcBasic and wrote a complete sw during christmas, at the same time I have put together hw for a part of the installation, and there I am. Totally crazy idea, but it keeps me be creative. Now you say "Phyton", never heard of it before I bought the Arduino, I will take a look at it. I have ordered a new usb-iic board (one chip and no internal comm), If it works, I am happy. Thank you for support.

**Reply by Marrin** [reply]
Python has a compiler, to byte code that is then interpreted, but it's transparent to the user.  There is not an extra compile step, this happens automatically when source files are run directly or imported from other source files/modules.  In case of imports the compiled byte code is usually cached so it does get re-compiled only when the source file changed.  And when you just start the Python executable you get an interactive prompt where you can enter Python code that gets evaluated directly after hitting the enter key.  Quite similar to PC-BASICs immediate mode, i.e. what happens when you enter statements without a line number in front, that are executed immediately.

If you don't want to switch the programming language completely and an actual explicit, mandatory compile step is tolerable, you may look into [FreeBASIC](https://freebasic.net/).  That started out as a free compiler for GW-BASICs successor QBasic/QuickBasic and at least compiles your example/test code from above with the `-lang qb` switch.  I can't actually test it.  But the `FOR XX…` delay loops will run much too fast compiled to machine code.

Although QBasic can run GW-BASIC code with almost no changes, it enables to write more readable and structured code.  For instance you don't need line numbers any more and `IF`/`ELSE` branches can span several lines.

```freebasic
OPEN "com1:9600,N,8,1,RS,CS0,DS0,CD0" FOR RANDOM AS #1
DO
  INPUT A
  IF A = 256 THEN
    REM * read *
    A$ = CHR$(&H53) + CHR$(&H41)
    PRINT #1, A$;
    SLEEP 1
    IF NOT EOF(1) THEN
      B$ = INPUT$(1, #1)
      PRINT "data in"
      PRINT VAL(B$)
    END IF
  ELSE
    REM * write to pcf 8574 *
    A$ = CHR$(&H53) + CHR$(&H40) + CHR$(A)
    PRINT #1, A$;
    SLEEP 1
    IF NOT EOF(1) THEN
      C$ = INPUT$(1, #1)
      PRINT "ack"
      PRINT VAL(C$)
    END IF
  END IF
LOOP
```

The top `IF` and `ELSE` branches end with nearly identical code which can be factored out into a named sub routine:

```freebasic
OPEN "com1:9600,N,8,1,RS,CS0,DS0,CD0" FOR RANDOM AS #1
DO
  INPUT A
  IF A = 256 THEN
    REM * read *
    A$ = CHR$(&H53) + CHR$(&H41)
    Communicate 1, A$, "data in"
  ELSE
    REM * write to pcf 8574 *
    A$ = CHR$(&H53) + CHR$(&H40) + CHR$(A)
    Communicate 1, A$, "ack"
  END IF
LOOP

SUB Communicate (f, data$, resultName$)
  PRINT #f, data$;
  SLEEP 1
  IF NOT EOF(f) THEN
    b$ = INPUT$(1, #f)
    PRINT resultName$
    PRINT VAL(b$)
  END IF
END SUB
```

**Reply by fakircurry** [reply]
Very interesting information. When I searched after "Basic for PC" recently, I found PC-Basic. It was exactly that I used long time ago. Now I understand that there are a lot of other similar alternatives. I will take a look at FreeBASIC and try to compile and test it on my hw.

**Comment by fakircurry** [comment]
Some information. After receiving the new USB to iic module everything works perfect. I have now written about 1000 lines of code and can comtrol my whole model train installation. All parts are controlled via a total of 160 relays and 24 inputs. Thanks for your help.
